### ISSUE TITLE
QC for {_inferred_1_M_13} ({BosTaurus})

### ISSUE THREAD

**Issue by iechevarriaz** [popsim-consortium/stdpopsim#579]
**PR for new model:** https://github.com/popsim-consortium/stdpopsim/pull/558
**Original paper:** https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3748359/
**Parameter values:** {reference to where parameter values can be found}
**Population indices:** {description of population indices}

**Potential issues that might lead to differences between the model implementations:**
{description of any potential issues}

**QC'er requests:** {tag potential developers to QC the model}

**Comment by gregorgorjanc** [comment]
Just a note here that I have "eyeballed" with @iechevarriaz the effective population sizes at different times from this paper https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3748359/ Figure 4b. These are the best estimates, from my knowledge.

**Comment by iechevarriaz** [comment]
Good morning! There are values for Ne and generations ago in the following supplementary material https://academic.oup.com/mbe/article/30/9/2209/1002512#supplementary-data

**Comment by agladstein** [comment]
I will work on this QC!

**Comment by grahamgower** [comment]
Any updates here @agladstein? It would be great to get Bos taurus in. Should we split the species and the demographic model into two commits?

**Comment by grahamgower** [comment]
I've taken a look at the demographic model while making pr #600, and it looks like the model implementation is not piecwise-constant, as in MacLeod et al. It's probably better to stick with the publication here, unless there's a good reason not to?

I.e. at the moment, the model has:
```
        demographic_events=[
            msprime.PopulationParametersChange(
                time=1,
                initial_size=90,
                growth_rate=-0.095894024,
                population_id=0,
            ),  # Ne 90 to 120
            msprime.PopulationParametersChange(
                time=4, growth_rate=-0.24465639, population_id=0
            ),  # Ne 120 to 250
            msprime.PopulationParametersChange(
                time=7, growth_rate=-0.0560787, population_id=0
            ),  # Ne 250 to 350
```

But MacLeod et al. actually shows something more like:
```
        demographic_events=[
            msprime.PopulationParametersChange(
                time=1,
                initial_size=90,
                growth_rate=0,
                population_id=0,
            ),  # Ne = 90
            msprime.PopulationParametersChange(
                time=4, initial_size=120, population_id=0
            ),  # Ne = 120
            msprime.PopulationParametersChange(
                time=7, initial_size=250, population_id=0
            ),  # Ne = 250
```

**Comment by gregorgorjanc** [comment]
Thanks @grahamgower! @iechevarriaz and I were not sure what would be best. In reality the change was likely gradual, but I see merit in exactly following the publication. Please advise.

**Comment by grahamgower** [comment]
Well using a growth rate like that will make each epoch have an exponential population size change, which is probably not what is desired either. So I would recommend just sticking with a piecewise-constant model. (Maybe in the future we'll try to include demographic models with nice smooth changes through time as in https://www.biorxiv.org/content/10.1101/2020.09.21.307355v1)

**Comment by gregorgorjanc** [comment]
OK, makes sense. This feedback is awesome.

**Comment by agladstein** [comment]
@grahamgower just to clarify, we still need to do a blind implementation for QC of this model right?

**Comment by grahamgower** [comment]
Yes please @agladstein! This will also need #601 and #606 before QC is complete. Maybe you'd prefer to work from the values in the parameter table (#606) and double check they match the figure in the paper?

**Comment by gregorgorjanc** [comment]
> Yes please @agladstein! This will also need #601 and #606 before QC is complete

#609 addresses the #601 and #606

**Comment by agladstein** [comment]
Great. I'll wait until https://github.com/popsim-consortium/stdpopsim/pull/609 gets merged.

**Comment by gregorgorjanc** [comment]
@agladstein #609 has been merged in. Maybe we do QC together at the "Grow the Zoo" hackathon?

**Comment by agladstein** [comment]
I was planning to add in a dog demographic model. 
Maybe there is time to do both?

**Comment by gregorgorjanc** [comment]
Purfect! You can QC my cattle species&model and I QC for your dog species&model.

**Comment by grahamgower** [comment]
Was there any progress on this QC @agladstein?

**Comment by igronau** [comment]
I can do this QC now. I will do this as a piecewise constant model using the values in Supplementary Table S1. Note that this corresponds to Fig 4A and not Fig 4B. See this text from page 5 in the SI:
> We therefore implemented forward-in-time simulations with FREGENE software (Hoggart, et al. 2007) using our final inferred Holstein demography (**Fig 4A and Table S1**), with a constant recombination rate of 1 x 10-8 and mutation rate of 9.4 x 10-9

**Comment by gregorgorjanc** [comment]
Thanks @igronau ! Yeah, supplementary table is where we took the last values from.

**Comment by igronau** [comment]
I implemented the QC model and received two inconsistencies that I'd like to check with you:

1. I used a mutation rate of 0.94e-8 and you used a mutation rate of 1.00e-8. I took my mutation rate from the description of the simulations used to generate data. I assume that this value creates a good fit between simulated and real data. Not sure how much of a difference this makes.
> We therefore implemented forward-in-time simulations with FREGENE software (Hoggart, et al. 2007) using our final inferred Holstein demography (Fig 4A and Table S1), with a constant recombination rate of 1 x 10-8 and mutation rate of **9.4 x 10-9** (resulting in an average heterozygosity of 9 x 10-4). 

2. The first epoch in your model is 4 generations long and in my version it's 3 generations long, as specified in Supp Table S1. Do you have any reason for adding that one generation?

Please let me know if you'd like to modify these parts in your model or think that they should be modified in the QC model.

**Comment by gregorgorjanc** [comment]
@igronau thanks for pointing this out - I think we simply rounded up the mutation rate, but I agree that best if we use the reported value, the same for the number of generations. Shall I make the edits in the PR?

**Comment by igronau** [comment]
I think so. After you update the model, I will re-run the QC and then add a PR with the QC code.

**Comment by igronau** [comment]
@gregorgorjanc Another option is that I apply these two changes to your model (mutation rate <-- 9.4e-9 and first epoch <-- 3 yrs) and assign you to review the PR before accepted. Let me know what you prefer

**Comment by gregorgorjanc** [comment]
@igronau I have now looked at this, including rereading the MacLeod2013 paper and supplement https://academic.oup.com/mbe/article/30/9/2209/1002512 - in inferring the demography they used mutation rate = recombination rate = 1 x 10-8, but true in their simulations to corroborate their estimate in real data they have used mutation rate of 9.4 x 10-9 (based on their sequence error correction and discussions etc). So, in a way using 1 x 10-8, as I did, is correct, BUT I have used values from the table that technically speaking correspond to simulation (though they say it matched real data). I am leaning towards changing to 9.4 x 10-9 since they suggest that rate relates to error free sequence data, which is what simulation gives. I will do this change so we have this done properly for the manuscript as well!

@igronau I am not entirely clear on the epochs. The table in supplement https://academic.oup.com/mbe/article/30/9/2209/1002512 states that the 1st epoch lasted 3 generations - so my understanding was that the code should be like this

```
population_configurations=[
            #      3 generations at    90,     1-     3
            msprime.PopulationConfiguration(
                initial_size=90, metadata=populations[0].asdict()
            )
        ],
        # Here 'time' should be in generation notation ie. how many
        # generations ago were that Ne (effective population size)
        demographic_events=[
            #      3 generations at   120,     4-     6
            msprime.PopulationParametersChange(
                time=4, initial_size=120, population_id=0
            ),
```

and not saying that the change was at time 3 - the 2nd epoch runs from 4 to 6 generations (3 generations), but I am happy to change this if you think this is wrong.

**Comment by grahamgower** [comment]
Hey @gregorgorjanc. Unless I misunderstand, I think @igronau is referring to your first epoch, which would be the one with `Ne=90`. Since time starts at `t=0`, this will apply to the time interval `[0,3)`, thus spanning 4 generations. I think this implies there's an off-by-one error for all of the demographic event time values.

**Comment by gregorgorjanc** [comment]
@grahamgower I see now - ok, then I need to shave off one generation everywhere. The joy of different counting systems;)

**Comment by gregorgorjanc** [comment]
Thanks @igronau and @grahamgower for all hand holding. I have now added 0-based counting in time epochs.

I have another question (not sure if I should open a new issue?) - the assumed recombination rate in the MacLeod2013 paper https://academic.oup.com/mbe/article/30/9/2209/1002512 for estimating the demography was 1 x 10-8, but in the BosTau species I have implemented the more recent estimate of 0.926 x 10-8 from MaEtAl2015 https://doi.org/10.1371/journal.pgen.1005387. What should we do about this discrepancy? Can we add recombination rate to the `stdpopsim.DemographicModel`? How can I help?

Thanks!

**Comment by grahamgower** [comment]
MacLeod's method is based on IBD tracts, right? So the recombination rate actually matters here for the inference of population sizes? That's a fairly small discrepancy though, and we don't have any infrastructure for including model-specific recombination rates anyway, so I'd say lets not worry too much.

**Comment by grahamgower** [comment]
Hey @igronau, it would be great if you have some time to revisit your QC model now. I guess it should match up now and we can close off this issue soon?

**Comment by gregorgorjanc** [comment]
> MacLeod's method is based on IBD tracts, right? So the recombination rate actually matters here for the inference of population sizes? That's a fairly small discrepancy though, and we don't have any infrastructure for including model-specific recombination rates anyway, so I'd say lets not worry too much.

Yes, IBD tract based, so indeed recombination matters for the estimation of Ne. I agree, that the difference is not huge, so not a major issue, but it would be good to add recombination rate in  `stdpopsim.DemographicModel` for completeness - I am sure there will be cases where the difference could be much bigger leading to conflicting mixing of parameters from quite different studies. I will open a separate issue for that.

**Comment by petrelharp** [comment]
Hi, @gregorgorjanc @igronau - what's the status here? Would you like help finishing this off?

**Comment by petrelharp** [comment]
Great - but, it still needs to be QC'ed, right? Maybe I'm just totally lost, but I'm not finding anything in `qc/BosTau` nor am I finding a PR with @igronau's implementation?

**Comment by igronau** [comment]
I started the QC process a while back, but after finding some issues, I waited for @gregorgorjanc to resolve them. I will finish the QC-ing now. I don't know if I'll have time to do this until next Monday, but it's on my TODO list.

**Comment by igronau** [comment]
This passes the QC now. The automatic tests point out a discrepancy between the mutation rate in this model (9.4 e-9) and the one mentioned in the docs (1e-8) in file docs/parameter_tables/BosTau/HolsteinFriesian_1M13.csv. Should I simply modify the csv file before I submit the Pull request, @gregorgorjanc ?

**Comment by petrelharp** [comment]
No need to wait to submit the PR - might as well let us look at it?

**Comment by igronau** [comment]
Okay. So I'll submit a PR after I make the change also in the doc table (HolsteinFriesian_1M13.csv) to make it consistent with the model. If @gregorgorjanc suggests that we do something else, we can fix this.

**Comment by igronau** [comment]
Pull request #1272

**Comment by grahamgower** [comment]
Closed by #1272.
